File: Data-Structures-and-Algorithms/0144-BinaryTreePreorderTraversal/preorderTraversal.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def preorderTraversal(self, root):
        ans = []
        def pTfn(node):
            if node:
                ans.append(node.val)
                pTfn(node.left)
                pTfn(node.right)
    
        pTfn(root)
        return ans

root  = TreeNode(0)
one   = TreeNode(1)
two   = TreeNode(2)
three = TreeNode(3)
four = TreeNode(4)
root.left = one
root.right = two
two.right = four
four.right = three
print(Solution().preorderTraversal(root))

# O(N) where N is the total number of nodes
# O(N) for worst case and O(log2(N)) for best case
# Preorder Traversal

File: Data-Structures-and-Algorithms/0144-BinaryTreePreorderTraversal/preorderTraversal.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

const preorderTraversal = function(root) {
    if (!root) { 
        return [] 
    };
    let stack = [(root)];
    let ans = [];
    while (stack.length) {
        node = stack.pop()
        ans.push(node.val)
        if (node.right) {
            stack.push(node.right)
        }
        if (node.left) {
            stack.push(node.left)
        }
    }
    return ans
};

let root = new TreeNode(0);
let one = new TreeNode(1);
let two = new TreeNode(2);
let three = new TreeNode(3);
root.left = one;
root.right = two;
two.right = three;
console.log(preorderTraversal(root))

// O(N) where N is the total number of nodes
// O(N) for worst case and O(log2(N)) for best case
// Preorder Traversal